### DISCUSSION TITLE
Requesting help in fixing an issue with respect to executing mysql query using laravel query builder

### DISCUSSION THREAD

**Discussion by Agilan3996** [laravel/framework#49280] [Q&A]
Hi Everyone,

We have a project which built using Laravel. We were facing one peculiar issue, where the query executed using DB facade/query builder is getting executed against the MySQL database  for very long time(> 500 secs), whereas the raw query generated from the same query object getting executed within 2 seconds directly in mysql client or even when the raw query is executed from the code using DB::select($rawQuery). We have compared the raw queries and both are exactly same.

We have lot of queries in our application which are built using DB facade/query builder, but we are not facing this issue elsewhere. In one particular flow alone we are facing this issue and we are clueless on why it is happening.

We couldnt find any related discussions or posts in any other sites as well regarding this behaviour. I could see posts or blogs on slowness of eloquent but not on DB facade/query builder.

Can someone throw us light on why this is happening?

Thanks in advance..!

**Comment by Agilan3996** [comment]
Attaching the raw query generated for reference.

`SELECT DISTINCT table_1.id
FROM `zrx`.`table_1`
INNER JOIN
  (SELECT MAX(uw.completed_date) AS completed_date
   FROM `zrx`.`table_1` AS `uw`
   INNER JOIN `syn`.`table_2` AS `r` ON `r`.`users_wave_id` = `uw`.`id`
   LEFT JOIN `zrx`.`table_3` ON `table_3`.`id` = `uw`.`id`
   LEFT JOIN `zrx`.`table_4` ON `table_4`.`id` = `uw`.`user_id`
   INNER JOIN `syn`.`segments` AS `s` ON `s`.`users_wave_id` = `uw`.`id`
   WHERE s.segment_rule_id = 19660
     AND uw.status IN (1)
     AND uw.wave_id IN (94327,
                        92895,
                        91801,
                        90663,
                        89571,
                        88066)
     AND r.qid =493608
     AND CONVERT_TZ(uw.completed_date, "+00:00", IF(table_3.time_zone IS NULL
                                                    OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) <= ADDTIME("2023-12-07 00:00:00", "23:59:59")
     AND CONVERT_TZ(uw.completed_date, "+00:00", IF(table_3.time_zone IS NULL
                                                    OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) >= ADDTIME("2017-11-16 00:00:00", "00:00:00")
   GROUP BY COALESCE(NULLIF(table_3.vendor_npi,''), NULLIF(table_4.npi,''), uw.user_id),
            `s`.`value`,
            `uw`.`wave_id`) AS responseLatestOrEarliest ON `responseLatestOrEarliest`.`completed_date` = `table_1`.`completed_date`
INNER JOIN `zrx`.`waves` ON `table_1`.`wave_id` = `waves`.`id`
LEFT JOIN `zrx`.`table_4` ON `table_1`.`user_id` = `table_4`.`id`
LEFT JOIN `zrx`.`table_3` ON `table_3`.`id` = `table_1`.`id`
LEFT JOIN `syn`.`table_2` AS `responses` ON `responses`.`users_wave_id` = `table_1`.`id`
LEFT JOIN `syn`.`segments` ON `segments`.`users_wave_id` = `table_1`.`id`
LEFT JOIN `syn`.`segment_rules` ON `segment_rules`.`id` = `segments`.`segment_rule_id`
WHERE table_1.wave_id IN (94327,
                              92895,
                              91801,
                              90663,
                              89571,
                              88066)
  AND table_1.status IN (1)
  AND `responses`.`qid` = '493608'
  AND `segment_rules`.`id` IN ('19660')
  AND 1 = 1
  AND CONVERT_TZ(table_1.completed_date, "+00:00", IF(table_3.time_zone IS NULL
                                                          OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) <= ADDTIME("2023-12-07 00:00:00", "23:59:59")
  AND CONVERT_TZ(table_1.completed_date, "+00:00", IF(table_3.time_zone IS NULL
                                                          OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) >= ADDTIME("2017-11-16 00:00:00", "00:00:00")`

**Reply by henzeb** [reply]
Maybe use codeblock, it's unreadable right now...

**Reply by Agilan3996** [reply]
@henzeb I have already updated the query in the code block only.

**Reply by shaedrich** [reply]
And what query does [Laravel generate in comparison](https://laravel.com/docs/10.x/queries#debugging)?

**Reply by Agilan3996** [reply]
@shaedrich  Please find the query generated by Laravel. Laravel version is 9.1 and used dd() to get the raw sql generated by Laravel. Got the qid as a binding as well.

`
SELECT DISTINCT `table_1`.`id`
FROM `zrx`.`table_1`
INNER JOIN
  (SELECT MAX(uw.completed_date) AS completed_date
   FROM `zrx`.`table_1` AS `uw`
   INNER JOIN `syn`.`table_2` AS `r` ON `r`.`users_wave_id` = `uw`.`id`
   LEFT JOIN `zrx`.`table_3` ON `table_3`.`id` = `uw`.`id`
   LEFT JOIN `zrx`.`table_4` ON `table_4`.`id` = `uw`.`user_id`
   WHERE uw.status IN (1)
     AND uw.wave_id IN (94327)
     AND r.qid =493608
     AND CONVERT_TZ(uw.completed_date, "+00:00", IF(table_3.time_zone IS NULL
                                                    OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) <= ADDTIME("2023-12-14 00:00:00", "23:59:59")
     AND CONVERT_TZ(uw.completed_date, "+00:00", IF(table_3.time_zone IS NULL
                                                    OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) >= ADDTIME("2023-11-02 00:00:00", "00:00:00")
   GROUP BY COALESCE(NULLIF(table_3.vendor_npi,''), NULLIF(table_4.npi,''), uw.user_id),
            `uw`.`wave_id`) AS responseLatestOrEarliest ON `responseLatestOrEarliest`.`completed_date` = `table_1`.`completed_date`
INNER JOIN `zrx`.`waves` ON `table_1`.`wave_id` = `waves`.`id`
LEFT JOIN `zrx`.`table_4` ON `table_1`.`user_id` = `table_4`.`id`
LEFT JOIN `zrx`.`table_3` ON `table_3`.`id` = `table_1`.`id`
LEFT JOIN `syn`.`table_2` AS `responses` ON `responses`.`users_wave_id` = `table_1`.`id`
WHERE table_1.wave_id IN (94327)
  AND table_1.status IN (1)
  AND `responses`.`qid` = ?
  AND CONVERT_TZ(table_1.completed_date, "+00:00", IF(table_3.time_zone IS NULL
                                                          OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) <= ADDTIME("2023-12-14 00:00:00", "23:59:59")
  AND CONVERT_TZ(table_1.completed_date, "+00:00", IF(table_3.time_zone IS NULL
                                                          OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) >= ADDTIME("2023-11-02 00:00:00", "00:00:00")
`

**Reply by rdgout** [reply]
Here you go, it's still pretty unreadable though.

```sql
SELECT DISTINCT
	table_1.id
FROM
	zrx.table_1
	INNER JOIN (
		SELECT
			MAX(uw.completed_date) AS completed_date
		FROM
			zrx.table_1ASuw
			INNER JOIN syn.table_2 AS r ON r.users_wave_id = uw.id
			LEFT JOIN zrx.table_3 ON table_3.id = uw.id
			LEFT JOIN zrx.table_4 ON table_4.id = uw.user_id
		WHERE
			uw.status IN(1)
			AND uw.wave_id IN(94327)
			AND r.qid = 493608
			AND CONVERT_TZ(uw.completed_date, "+00:00", IF(table_3.time_zone IS NULL OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) <= ADDTIME("2023-12-14 00:00:00", "23:59:59") AND CONVERT_TZ(uw.completed_date, "+00:00", IF(table_3.time_zone IS NULL OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) >= ADDTIME("2023-11-02 00:00:00", "00:00:00")
		GROUP BY
			COALESCE(NULLIF(table_3.vendor_npi, ''), NULLIF(table_4.npi, ''), uw.user_id),
			uw.wave_id) AS responseLatestOrEarliest ON responseLatestOrEarliest.completed_date = table_1.completed_date
	INNER JOIN zrx.wavesONtable_1.wave_id = waves.id
	LEFT JOIN zrx.table_4 ON table_1.user_id = table_4.id
	LEFT JOIN zrx.table_3 ON table_3.id = table_1.id
	LEFT JOIN syn.table_2ASresponses ON responses.users_wave_id = table_1.id
WHERE
	table_1.wave_id IN(94327) AND table_1.status IN(1) AND responses.qid = ? AND CONVERT_TZ(table_1.completed_date, "+00:00", IF(table_3.time_zone IS NULL OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) <= ADDTIME("2023-12-14 00:00:00", "23:59:59") AND CONVERT_TZ(table_1.completed_date, "+00:00", IF(table_3.time_zone IS NULL OR table_3.time_zone = "-99.99", "+00:00", table_3.time_zone)) >= ADDTIME("2023-11-02 00:00:00", "00:00:00")
```

**Reply by Agilan3996** [reply]
Thanks for the help in making it readable @rdgout

**Reply by macropay-solutions** [reply]
Just a thought. When raw querying the eager loading is not happening. When executing the query using a builder maybe other stuff is executed also. I see you use distinct. The distinct has a fragile use case in laravel that if you don't get right, it backfires.

On a second thought, your initial sql ran via builder is cached in mysql and the second time you run it via raw query is very fast.
